ql/src/java/org/apache/hadoop/hive/ql/exec/repl/bootstrap/load/table/LoadTable.java

```diff
@@ -218,9 +219,25 @@ private String location(ImportTableDesc tblDesc, Database parentDb)
   private Task<?> loadTableTask(Table table, ReplicationSpec replicationSpec, Path tgtPath,
       Path fromURI) {
     Path dataPath = new Path(fromURI, EximUtil.DATA_PATH_NAME);
-    Path tmpPath = PathUtils.getExternalTmpPath(tgtPath, context.pathInfo);
-    Task<?> copyTask =
-        ReplCopyTask.getLoadCopyTask(replicationSpec, dataPath, tmpPath, context.hiveConf);
+    Path tmpPath = tgtPath;
+
+    // if move optimization is enabled, copy the files directly to the target path. No need to create the staging dir.
+    LoadFileType loadFileType;
+    if (replicationSpec.isInReplicationScope() &&
+            context.hiveConf.getBoolVar(REPL_ENABLE_MOVE_OPTIMIZATION)) {
+      loadFileType = LoadFileType.IGNORE;
+    } else {
+      loadFileType =
+              replicationSpec.isReplace() ? LoadFileType.REPLACE_ALL : LoadFileType.OVERWRITE_EXISTING;
+      tmpPath = PathUtils.getExternalTmpPath(tgtPath, context.pathInfo);
+    }
+
+    LOG.debug("adding dependent CopyWork/AddPart/MoveWork for table "
+            + table.getCompleteName() + " with source location: "
+            + dataPath.toString() + " and target location " + tmpPath.toString());
+
+    Task<?> copyTask = ReplCopyTask.getLoadCopyTask(replicationSpec, dataPath, tmpPath, context.hiveConf,
+            false, false);
```

Comment:
Check needCmRecycle flag.

Reply:
do we need to do cm recycle during boot strap ?